File: Algorithms_medium/34_Find First and Last Position of Element in Sorted Array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1, -1]
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) >> 1
            if nums[mid] >= target:
                right = mid
            else:
                left = mid + 1
        if nums[left] != target:
            return [-1, -1]
        l, right = left, len(nums) - 1
        while left < right:
            mid = (left + right + 1) >> 1
            if nums[mid] <= target:
                left = mid
            else:
                right = mid - 1
        return [l, left]
'''
用两次二分查找寻找左边界和右边界
主意找到的最初边界要等于寻找数，而不是在一个范围内
'''

File: Algorithms_C++/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, bool lower) {
        int left = 0, right = (int)nums.size()-1, ans = (int)nums.size();
        while (left <= right){
            int mid = (left + right) >> 1;
            if (nums[mid] > target || (lower && nums[mid] >= target)){
                right = mid - 1;
                ans = mid;
            }
            else{
                left = mid + 1;
            }
        }
        return ans;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = binarySearch(nums, target, true);
        int last = binarySearch(nums, target, false) - 1;
        if (first <= last && last < nums.size() && nums[first] == target && nums[last] == target){
            return vector<int> {first, last};
        }
        return vector<int> {-1, -1};
    }
};